modules/core/src/main/java/org/apache/ignite/internal/cdc/CdcMain.java

```diff
@@ -459,8 +472,33 @@ public void consumeWalSegmentsUntilStopped() {
                             }
 
                             lastSgmnt.set(nextSgmnt);
-                        })
-                        .forEach(this::consumeSegment); // Consuming segments.
+                        }).iterator();
+
+                    while (it.hasNext()) {
+                        Path segment = it.next();
+
+                        if (log.isInfoEnabled()) {
+                            log.info("Processing WAL segment [mode=" + (cdcMgrModeState ? "manager" : "file")
+                                + ", segment=" + segment + ']');
+                        }
+
+                        lastSegmentConsumptionTs.value(System.currentTimeMillis());
+                        curSegmentIdx.value(segmentIndex(segment));
+
+                        if (cdcMgrModeState) {
+                            consumeSegmentPassively(segment);
```

Comment:
```java
                            cdcMgrModeState = consumeSegmentPassively(segment);
                            if (cdcMgrModeState) {
                                processedSegments.add(segment);

                                clearSegments((walState = state.loadWalState()));
                            }
                            else {
                                // Reset passively handled segments handled. Need to consume it again.
                                seen.clear();
                                lastSgmnt.set(-1);

                                break;
                            }
```

Reply:
I reverted this, and use another WALRecord for sending walState instead of writing directly to the file. There are some issues when saved walState greater than a processing or known segment. It's pretty tricky to handle it within CdcMain. I suppose to use record is much more convenient and straightforward.